pydy/codegen/ode_function_generator.py

```diff
@@ -0,0 +1,705 @@
+
+    def create_rhs_function(self):
+
+        xdot = np.empty(self.num_states, dtype=float)
+
+        def rhs(*args):
+            # args: x, t, p
+            # args: x, t, r, p
+
+            args = self._parse_old_style_extra_args(*args)
+
+            args = self._parse_constants(*args)
+
+            if self.specifieds is not None:
+                args = self._parse_specifieds(*args)
+
+            q = args[0][:self.num_coordinates]
+            u = args[0][self.num_coordinates:]
+
+            M, F, qdot = self.eval_arrays(q, u, *args[2:])
+
+            if self.num_speeds == 1:
+                udot = F / M
+            else:
+                udot = self.solve_linear_system(M, F)
+
+            xdot[:self.num_coordinates] = qdot
+            xdot[self.num_coordinates:] = udot
+
+            return xdot
+
+        rhs.__doc__ = self._generate_rhs_docstring()
+
+        return rhs
+
+
+class ODEFunctionGenerator(object):
+
+    _rhs_doc_template = \
+"""\
+Returns the derivatives of the states, i.e. numerically evaluates the right
+hand side of the first order differential equation.
+
+x' = f(x, t,{specified_call_sig} p)
+
+Parameters
+==========
+x : ndarray, shape({num_states},)
+    The state vector is ordered as such:
+{state_list}
+t : float
+    The current time.{specified_explanation}
+p : dictionary len({num_constants}) or ndarray shape({num_constants},)
+    The dictionary maps the constant symbols to floats. The constants are,
+    in order:
+{constant_list}
+
+Returns
+=======
+dx : ndarray, shape({num_states},)
+    The derivative of the state vector.
+
+"""
+
+    _specifieds_doc_template = \
+"""
+r : dictionary; ndarray, shape({num_specified},); function
+
+    There are three options for this dictionary. (1) is more flexible but
+    (2) and (3) are much more efficient.
+
+    (1) A dictionary that maps the specified functions of time to floats,
+    ndarrays, or functions that produce ndarrays. The keys can be a single
+    specified symbolic function of time or a tuple of symbols. The total
+    number of symbols must be equal to {num_specified}. If the value is a
+    function it must be of the form g(x, t), where x is the current state
+    vector ndarray and t is the current time float and it must return an
+    ndarray of the correct shape. For example::
+
+      r = {{a: 1.0,
+           (d, b) : np.array([1.0, 2.0]),
+           (e, f) : lambda x, t: np.array(x[0], x[1]),
+           c: lambda x, t: np.array(x[2])}}
+
+    (2) A ndarray with the specified values in the correct order and of the
+    correct shape.
+
+    (3) A function that must be of the form g(x, t), where x is the current
+    state vector and t is the current time and it must return an ndarray of
+    the correct shape.
+
+    The specified inputs are, in order:
+{specified_list}\
+"""
+
+    @staticmethod
+    def _deduce_system_type(**kwargs):
+
+        if kwargs.pop('coordinate_derivatives') is not None:
```

Comment:
I'm a little confused by this. Are these user set kwargs? They appear to work as long as their defined, so I could pass in `coordinate_derivatives=True` or `coordinate_derivatives=False`, or `coordinate_derivatives="yay duck typing"`, and get the same result each time?

What is the meaning of the different options? Can they contradict each other?